planning.py

```diff
@@ -1435,3 +1587,105 @@ def go_to_sfo():
+            clauses = conjuncts(clauses)
+            for i in range(len(clauses)):
+                for ch in lib.keys():
+                    if clauses[i].op == ch:
+                        clauses[i] = expr( lib[ch] + str(clauses[i].args[0]))
+
+        elif isinstance(clauses, str):
+            for ch in lib.keys():
+                clauses = clauses.replace(ch, lib[ch])
+            if len(clauses) > 0:
+                clauses = expr(clauses)
+
+            try:
+                clauses = conjuncts(clauses)
+            except AttributeError:
+                pass
+
+        return clauses
+
+        
+        
+
+    def angelic_action(self):
+        """
+        example: 
+        """
+        #print(self.effect)
+
+        effects=[[]]
+        for clause in self.effect:
+            if clause.op[:9] == 'PosYesNot':
+                effects = effects*3
+                n=3
+                w=9
+            elif clause.op[:6] == 'PosYes':
+                effects =effects* 2
+                n=2
+                w=6
+            elif clause.op[:6] == 'PosNot':
+                effects = effects* 2
+                n=2
+                w=3
+            else:
+                n=1
+                w=0
+            it=range(1)
+            if len(effects)!=0:
+                it = range(len(effects)//n)
+            for i in it:
+                if effects[i]:
+                    if clause.args: 
+                        effects[i] = expr(str(effects[i]) + '&' + str(Expr(clause.op[w:],clause.args[0])))
+                        if n==3:
+                            effects[i+len(effects)//3]= expr(str(effects[i+len(effects)//3]) + '&' + str(Expr(clause.op[6:],clause.args[0])))
+                    else: 
+                        effects[i] = expr(str(effects[i]) + '&' + str(expr(clause.op[w:])))
+                        if n==3: effects[i+len(effects)//3] = expr(str(effects[i+len(effects)//3]) + '&' + str(expr(clause.op[6:])))
+
+                else: 
+                    if clause.args: 
+                        effects[i] = Expr(clause.op[w:], clause.args[0])
+                        if n==3: effects[i+len(effects)//3] = Expr(clause.op[6:], clause.args[0])
+
+                    else: 
+                        effects[i] = expr(clause.op[w:])
+                        if n==3: effects[i+len(effects)//3] = expr(clause.op[6:])
+            #print('effects',  effects)
+        
+        return [ HLA(Expr(self.name, self.args), self.precond, effects[i] ) for i in range(len(effects)) ]
+
+
+class Angelic_Node(Node):
+    """ 
+    Extends the class Node. 
+    self.action:     contains the optimistic description of an angelic HLA
+    self.action_pes: contains the pessimistic description of an angelic HLA
+    """
+
+    def __init__(self, state, parent=None, action_opt=None, action_pes=None, path_cost=0):
+        super().__init__(state, parent, action_opt, path_cost)
+        self.action_pes = action_pes 
+
```

Comment:
Why is action_opt being set by the parent init method and action_pes being set by this init method, instead of both of them being set in this init method? Just curious.

Reply:
I just wanted to make full use of the parent object's variables, so I used the action variable to represent the optimistic description of the action. 
It is very easy for me to change it and add a variable action_opt.